Provide suggestions for type parameters missing bounds for associated types

estebank opened the pull request:
When implementing the binary operator traits it is easy to forget to restrict the `Output` associated type. `rustc` now accounts for different cases to lead users in the right direction to add the necessary restrictions. The structured suggestions in the following output are new:

```
error: equality constraints are not yet supported in `where` clauses
  --> $DIR/missing-bounds.rs:37:33
   |
LL | impl<B: Add> Add for E<B> where <B as Add>::Output = B {
   |                                 ^^^^^^^^^^^^^^^^^^^^^^ not supported
   |
   = note: see issue #20041 <https://github.com/rust-lang/rust/issues/20041> for more information
help: if `Output` is an associated type you're trying to set, use the associated type binding syntax
   |
LL | impl<B: Add> Add for E<B> where B: Add<Output = B> {
   |                                 ^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
  --> $DIR/missing-bounds.rs:11:11
   |
7  | impl<B> Add for A<B> where B: Add {
   |      - this type parameter
...
11 |         A(self.0 + rhs.0)
   |           ^^^^^^^^^^^^^^ expected type parameter `B`, found associated type
   |
   = note: expected type parameter `B`
             found associated type `<B as std::ops::Add>::Output`
help: consider further restricting this bound
   |
7  | impl<B> Add for A<B> where B: Add + std::ops::Add<Output = B> {
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0369]: cannot add `B` to `B`
  --> $DIR/missing-bounds.rs:31:21
   |
31 |         Self(self.0 + rhs.0)
   |              ------ ^ ----- B
   |              |
   |              B
   |
help: consider restricting type parameter `B`
   |
27 | impl<B: std::ops::Add<Output = B>> Add for D<B> {
   |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

That output is given for the following cases:

```rust
struct A<B>(B);
impl<B> Add for A<B> where B: Add {
    type Output = Self;

    fn add(self, rhs: Self) -> Self {
        A(self.0 + rhs.0) //~ ERROR mismatched types
    }
}

struct D<B>(B);
impl<B> Add for D<B> {
    type Output = Self;

    fn add(self, rhs: Self) -> Self {
        Self(self.0 + rhs.0) //~ ERROR cannot add `B` to `B`
    }
}

struct E<B>(B);
impl<B: Add> Add for E<B> where <B as Add>::Output = B {
    type Output = Self;

    fn add(self, rhs: Self) -> Self {
        Self(self.0 + rhs.0)
    }
}
```

estebank committed:
Handle binop on unbound type param

When encountering a binary operation involving a type parameter that has
no bindings, suggest adding the appropriate bound.

estebank committed:
Suggest restricting type param when it doesn't satisfy projection

When encountering a projection that isn't satisfied by a type parameter,
suggest constraining the type parameter.

rust-highfive commented:
r? @matthewjasper

(rust_highfive has picked a reviewer for you, use r? to override)

Centril commented on src/librustc_hir/hir.rs:
```diff
@@ -2648,8 +2648,38 @@ impl Node<'_> {
         match self {
             Node::TraitItem(TraitItem { generics, .. })
             | Node::ImplItem(ImplItem { generics, .. })
+            | Node::Item(Item { kind: ItemKind::Trait(_, _, generics, ..), .. })
+            | Node::Item(Item { kind: ItemKind::Impl { generics, .. }, .. })
             | Node::Item(Item { kind: ItemKind::Fn(_, generics, _), .. }) => Some(generics),
```
Can use an or-pattern in `kind:`

estebank committed:
review comments: use or-pattern

estebank committed:
On incorrect equality constraint likely to be assoc type, suggest appropriate syntax

When encountering `where <A as Foo>::Bar = B`, it is possible that `Bar`
is an associated type. If so, suggest `where A: Foo<Bar = B>`.

CC #20041.

estebank commented on src/test/ui/generic-associated-types/construct_with_other_type.stderr:
```diff
@@ -2,11 +2,16 @@ error[E0271]: type mismatch resolving `for<'a> <<T as Baz>::Baa<'a> as std::ops:
    |
 LL | impl<T> Baz for T where T: Foo {
-   |         ^^^ expected type parameter `T`, found associated type
+   |      -  ^^^ expected type parameter `T`, found associated type
+   |      |
+   |      this type parameter
    |
    = note: expected associated type `<T as Foo>::Bar<'_, 'static>`
               found associated type `<<T as Baz>::Quux<'_> as Foo>::Bar<'_, 'static>`
-   = note: you might be missing a type parameter or trait bound
+help: consider further restricting this bound
+   |
+LL | impl<T> Baz for T where T: Foo + Baz<Quux = T> {
+   |                                ^^^^^^^^^^^^^^^
```
This suggestion will  cause another error:

```
error[E0275]: overflow evaluating the requirement `<T as Baz>::Quux`
  --> src/test/ui/generic-associated-types/construct_with_other_type.rs:19:9
   |
19 | impl<T> Baz for T where T: Foo + Baz<Quux = T> {
   |         ^^^
   |
   = note: required because of the requirements on the impl of `Baz` for `T`
   = note: required because of the requirements on the impl of `Baz` for `T`
```

I feel this is "fine" since the output was _already_ confusing and the code itself is a bit contrived and requires `#![feature(generic_associated_types)]` to be enabled to even be triggered.

estebank commented:
For tracking purposes, the original impetus to make this change came from seeing this thread https://users.rust-lang.org/t/you-might-be-missing-a-type-parameter-or-trait-bound/38333.

This PR should help

- newcomers that want to apply a binary operation to two type arguments
- people trying to preempt the missing bound error without knowing the whole trait definition
- people who know that they want to specify the associated type but are unsure about the syntax and follow the code that unmet predicate errors show

These three cases are handled by different code (in different commits), but _all_ should be handled in concert to catch people at any point of their "incorrect mental model" trip. The last one is particularly insidious because the associated type binding equality constraint syntax _can_ be extrapolated from the rest of the syntax and displayed errors and _is_ being worked on, but is not yet available. Pointing people in the right direction instead of just saying "sorry, computer says no" is important to the learnability of the language.

nikomatsakis commented:
r? @nikomatsakis

nikomatsakis commented on src/librustc_typeck/check/op.rs:
```diff
@@ -253,6 +254,50 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                            let hir = self.tcx.hir();
+                            let msg =
+                                &format!("`{}` might need a bound for `{}`", lhs_ty, missing_trait);
+                            if let Some(def_id) = hir
```
Nit: I think a comment would be nice here. I think something like 

---

Try to find the def-id and details for the parameter `p`. We have only the index,
so we have to find the enclosing function's def-id, then look through its declared
generic parameters to get the declaration.

---

Come to think of it, though, we have the `HirId` of the *body* being checked, would it be easier to invoke [`body_owner_def_id`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/map/struct.Map.html#method.body_owner_def_id) on `self.body_id`?

estebank replied:
I used `body_id`, but it is a `hir::HirId`. I only wrapped it with a ` hir::BodyId` but a slightly larger change would be to change `body_id` to be a `BodyId` instead, but I didn't want to bloat this PR for the reviewer's sake :)

nikomatsakis commented on src/librustc_middle/ty/diagnostics.rs:
```diff
@@ -63,3 +68,180 @@ impl<'tcx> TyS<'tcx> {
+        );
+    };
+
+    if param_name.starts_with("impl ") {
```
Oof, hacky. =) But I don't have a better suggestion.

estebank replied:
This hack _is_ my fault, but preexisting to this PR 😅

nikomatsakis reviewed (commented):
Didn't get all the way through review this yet, just the first commit. Left some comments.

bors commented:
:umbrella: The latest upstream changes (presumably #71201) made this pull request unmergeable. Please resolve the merge conflicts.

estebank committed:
review comments and rebase fix

estebank committed:
review comment: use `body_id`

bors commented:
:umbrella: The latest upstream changes (presumably #71410) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented on src/librustc_ast_passes/ast_validation.rs:
```diff
@@ -1300,6 +1291,87 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
     }
 }
 
+fn deny_equality_constraints(
```
It'd be nice to have a brief comment

nikomatsakis commented on src/librustc_middle/ty/diagnostics.rs:
```diff
@@ -67,3 +72,180 @@ impl<'tcx> TyS<'tcx> {
     }
 }
+
+/// Suggest restricting a type param with a new bound.
```
It'd be nice to add a brief example.

nikomatsakis commented on src/librustc_middle/ty/diagnostics.rs:
```diff
@@ -67,3 +72,180 @@ impl<'tcx> TyS<'tcx> {
+    if generics.where_clause.predicates.is_empty()
+        // Given `trait Base<T = String>: Super<T>` where `T: Copy`, suggest restricting in the
+        // `where` clause instead of `trait Base<T: Copy = String>: Super<T>`.
+        && !matches!(param.kind, hir::GenericParamKind::Type { default: Some(_), .. })
```
```
!matches!
```

what have we wrought

estebank replied:
Clearly this should be `¡Mátches!`.

nikomatsakis commented on src/librustc_middle/ty/diagnostics.rs:
```diff
@@ -67,3 +72,180 @@ impl<'tcx> TyS<'tcx> {
+
+    if generics.where_clause.predicates.is_empty()
+        // Given `trait Base<T = String>: Super<T>` where `T: Copy`, suggest restricting in the
+        // `where` clause instead of `trait Base<T: Copy = String>: Super<T>`.
```
Is this comment accurate? It seems like this covers the case where we *don't* suggest a where clause, at least some of the time

nikomatsakis reviewed (approved):
I left a few nits but they don't matter really

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit c0a9d52b6f870d7401639e8a1ae8f2def6828586 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis c0a9d52b6f870d7401639e8a1ae8f2def6828586 -->
<!-- homu: {"type":"Approved","sha":"c0a9d52b6f870d7401639e8a1ae8f2def6828586","approver":"nikomatsakis"} -->

Dylan-DPC commented:
failed in rollup

@bors r-

bors commented:
:umbrella: The latest upstream changes (presumably #71108) made this pull request unmergeable. Please resolve the merge conflicts.

estebank committed:
fix rebase

estebank committed:
Add docstring to `deny_equality_constraints`

estebank commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit b17b20cafc87193dace2beb1cdb0e126d944f8e3 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis b17b20cafc87193dace2beb1cdb0e126d944f8e3 -->
<!-- homu: {"type":"Approved","sha":"b17b20cafc87193dace2beb1cdb0e126d944f8e3","approver":"nikomatsakis"} -->

bors commented:
:hourglass: Testing commit b17b20cafc87193dace2beb1cdb0e126d944f8e3 with merge 7ab8af3dd0318629a0a3d835468d9e3eb121a948...
<!-- homu: {"type":"BuildStarted","head_sha":"b17b20cafc87193dace2beb1cdb0e126d944f8e3","merge_sha":"7ab8af3dd0318629a0a3d835468d9e3eb121a948"} -->

Dylan-DPC commented:
@bors retry yied
